pandas/io/excel/_openpyxl.py

```diff
@@ -412,7 +419,25 @@ def write_cells(
         _style_cache: Dict[str, Dict[str, Serialisable]] = {}
 
         if sheet_name in self.sheets:
-            wks = self.sheets[sheet_name]
+            if "r+" in self.mode:
+                if self.if_sheet_exists == "new":
+                    wks = self.book.create_sheet()
+                    # openpyxl will create a name for the new sheet by appending digits
```

Comment:
This is all handled by openpyxl (https://foss.heptapod.net/openpyxl/openpyxl/-/blob/branch/3.0/openpyxl/workbook/child.py#L15). Seems unlikely that this will change so I'll add a note to the docs explaining it. If append mode is implemented for any other engines then this will probably be different

Reply:
Ah, good point. Then just an indication that the new name is determined by the engine used I think.